opentracing/src/main/java/opentracing/TraceContextDecoder.java

```diff
@@ -0,0 +1,41 @@
+/**
+ * Copyright 2016 The OpenTracing Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package opentracing;
+
+import java.util.Map;
+
+/**
+ * Decodes a trace context from binary or text formats.
+ *
+ * @see TraceContextEncoder
+ */
+// See https://github.com/opentracing/opentracing.github.io/issues/21 about folding codec together
+public interface TraceContextDecoder {
```

Comment:
TL;DR: the oddities in the coding API grew out of the following:
- a requirement to avoid cruft and/or unnecessary copies of TraceContext representation data (not of the trace attributes as much as whatever "context ids" are intrinsic to the particular TraceContext impl... trace_id, span_id, that sort of thing)... i.e., if a TraceContext knows that it only has a trace_id and span_id, it can encode those positionally rather than as a map.
- to that end, the context identifiers are often bounded and – per a comment @yurishkuro made a little while back – sometimes encoded in a fixed-length buffer; the trace attributes are not, hence the way the go and python APIs separate those two pieces of the encoded `TraceContext` (*)
- re supporting text and binary but nothing else: TBH, I would probably have just gone for binary and told folks that wanted to stuff them in HTTP headers to base64 encode or whatever. But my understanding is that many zipkin users really like having the various fields split out in human-readable (and software-router-accessible) fashion in HTTP headers, hence the text format. We hope (?) that there will only be two, hence cramming them into one interface.

Now, presently the encoder and decoder are separate interfaces... there is an issue open to change that and I haven't heard about any opposition. https://github.com/opentracing/opentracing.github.io/issues/21

(*): @adriancole I just noticed that you are encoding/decoding to/from a *single* `byte[]` and a *single* `Map<...>`, not two like py and go. Is that deliberate? If not, let's change it; if so, I'm curious about the semantic divergence.